File: backend/app/schemas/recipe.py
from typing import Literal

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel

MealType = Literal["breakfast", "lunch", "dinner", "snack"]
SafetyRating = Literal["safe", "caution", "avoid"]
IngredientCategory = Literal["proteins", "grains", "vegetables", "fruits", "essentials", "other"]


class _CamelModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True, from_attributes=True)


class IngredientOut(_CamelModel):
    id: int
    item: str
    amount: str
    category: IngredientCategory
    naira_kobo: int
    sort_order: int


class StepOut(_CamelModel):
    id: int
    step_number: int
    instruction: str


class RecipeSummary(_CamelModel):
    id: int
    slug: str
    name: str
    local_name: str | None
    meal_type: MealType
    image_url: str
    portion: str
    reason: str
    calories: int
    protein_g: int
    carbs_g: int
    fat_g: int
    sodium_mg: int
    potassium_mg: int
    prep_minutes: int
    cook_minutes: int
    servings: int
    cuisine_region: str | None
    condition_safety: dict[str, SafetyRating]


class RecipeDetail(RecipeSummary):
    description: str | None
    ingredients: list[IngredientOut]
    steps: list[StepOut]

File: frontend/src/types/recipe.ts
export type MealType = "breakfast" | "lunch" | "dinner" | "snack";
export type SafetyRating = "safe" | "caution" | "avoid";
export type ConditionSafety = Record<string, SafetyRating>;

export type IngredientCategory =
  | "proteins"
  | "grains"
  | "vegetables"
  | "fruits"
  | "essentials"
  | "other";

export interface Ingredient {
  id: number;
  item: string;
  amount: string;
  category: IngredientCategory;
  nairaKobo: number;
  sortOrder: number;
}

export interface RecipeStep {
  id: number;
  stepNumber: number;
  instruction: string;
}

export interface RecipeSummary {
  id: number;
  slug: string;
  name: string;
  localName: string | null;
  mealType: MealType;
  imageUrl: string;
  portion: string;
  reason: string;
  calories: number;
  proteinG: number;
  carbsG: number;
  fatG: number;
  sodiumMg: number;
  potassiumMg: number;
  prepMinutes: number;
  cookMinutes: number;
  servings: number;
  cuisineRegion: string | null;
  conditionSafety: ConditionSafety;
}

export interface Recipe extends RecipeSummary {
  description: string | null;
  ingredients: Ingredient[];
  steps: RecipeStep[];
}
